### PR TITLE
PixelShuffle check that output is not null before applying kernel (#85155)

### LINKED ISSUE #85155: torch.nn.PixelShuffle crash with floating point exception when input has 0 size in the last three dimensions
### 🐛 Describe the bug

`torch.nn.PixelShuffle` crash with floating point exception when input has 0 size in the last three dimensions

```
import torch
layer = torch.nn.PixelShuffle(1)

model_input = torch.ones((1,1,1,1,0))
pred = layer(model_input)
```

Also crash when input is 
```
model_input = torch.ones((1,1,1,1,0))
model_input = torch.ones((1,1,1,0,1))
model_input = torch.ones((1,1,0,1,1))
```

Output:
```
Floating point exception (core dumped)
```

### Versions

Collecting environment information...
PyTorch version: 1.13.0.dev20220915+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.10

Python version: 3.7.6 (default, Jan  8 2020, 19:59:22)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.15.0-176-generic-x86_64-with-debian-buster-sid
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.13.0.dev20220915+cpu
[pip3] torchaudio==0.13.0.dev20220915+cpu
[pip3] torchvision==0.14.0.dev20220915+cpu
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py37he8ac12f_0  
[conda] mkl_fft                   1.3.0            py37h54f3939_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.19.2           py37h54aff64_0  
[conda] numpy-base                1.19.2           py37hfa32c7d_0  
[conda] torch                     1.13.0.dev20220915+cpu          pypi_0    pypi
[conda] torchaudio                0.13.0.dev20220915+cpu          pypi_0    pypi
[conda] torchvision               0.14.0.dev20220915+cpu          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by Thytu**
Checks that tensor is not null before applying kernel in `pixel_shuffle` op

Signed-off-by: Thytu <valentin.de-matos@epitech.eu>

Fixes #85155

### CODE REVIEW

**Comment by lezcano** [comment]
Yes, add a test please.

`math_pixed_shuffle` is used in any dispatch that's not CPU. For example, CUDA, XLA, META, LTC (LazyTensor Core)...

**Comment by Thytu** [comment]
> 

Roger that

**Comment by Thytu** [comment]
Test added but not sure if I it's the right place to put it as a `FunctionalTest` (it's rather a UT IMO) but didn't find another place to test in the C++ API.

Would you prefer the test to be in `test/test_nn.py` ?

File: test/cpp/api/functional.cpp [hunk 1]
```diff
@@ -1670,6 +1670,21 @@ TEST_F(FunctionalTest, PixelShuffle) {
...
+  auto y_zero_dim1 = F::pixel_shuffle(x_zero_dim1, 1);
+
+  auto x_zero_dim2 = torch::ones({1,1,0,1,1}, torch::kFloat);
+  auto y_zero_dim2 = F::pixel_shuffle(x_zero_dim2, 1);
+
+  ASSERT_TRUE(y_zero_dim0.numel() == 0);
+  ASSERT_TRUE(y_zero_dim1.numel() == 0);
+  ASSERT_TRUE(y_zero_dim2.numel() == 0);
+}
+
```
Comment (lezcano) [inline test/cpp/api/functional.cpp:1687]: May be better to test through the Python API,  adding some inputs to the OpInfo of `nn.functional.pixel_shuffle`

Comment (Thytu) [reply test/cpp/api/functional.cpp:1687]: roger that!

Comment (Thytu) [reply test/cpp/api/functional.cpp]: changed in bdda9b6f8f1e2ac09749d9fea08c109ce72555aa

**Comment by Thytu** [comment]
It crashs here when running [test_ops_gradients.py](https://github.com/pytorch/pytorch/blob/8bad17f6b97d719a7dcece99ac91126a3029a10b/test/test_ops_gradients.py#L20) : https://github.com/pytorch/pytorch/blob/8bad17f6b97d719a7dcece99ac91126a3029a10b/torch/autograd/__init__.py#L300-L302

Still inspecting

**Comment by lezcano** [comment]
As per https://github.com/pytorch/pytorch/issues/85155#issuecomment-1250791994, you added the check to just one of the two functions and not the other, so when the other one is being executed it segfaults :D

**Comment by Thytu** [comment]
> As per [#85155 (comment)](https://github.com/pytorch/pytorch/issues/85155#issuecomment-1250791994), you added the check to just one of the two functions and not the other, so when the other one is being executed it segfaults :D

I thought the same but I'm not sure as:
- I'm testing by running `python -m pytest test/test_ops_gradients.py -k test_fn_fwgrad_bwgrad_nn_functional_pixel_shuffle_cpu_float64` (so it runs on CPU)

- I force check that it's not caused by `math_pixel_shuffle` by changing it to :
```cpp
Tensor math_pixel_shuffle(const Tensor& self, int64_t upscale_factor) {
  return self;
```

- The following code works :
```py
torch.nn.functional.pixel_shuffle(
    input=torch.ones((1, 1, 1, 0), dtype=torch.float64).to(torch.device('cuda')),
    upscale_factor=1,
)
```

**Comment by mruberry** [comment]
The new sample inputs look like the right coverage but I think the crashes just need more debugging to understand.

**Comment by lezcano** [comment]
It may be that the backward for this function also needs a similar treatment. The equivalent of `native_functions.yaml` is called `derivatves.yaml`.

**Comment by Thytu** [comment]
> It may be that the backward for this function also needs a similar treatment. The equivalent of `native_functions.yaml` is called `derivatives.yaml`.

Alright I will take a look at it, thanks

**Comment by Thytu** [comment]
> It may be that the backward for this function also needs a similar treatment. The equivalent of `native_functions.yaml` is called `derivatives.yaml`.

As it's an `auto_linear` it's autogenerated.

Quick question :what would the grad supposed to be as the output is empty? `0`?

**Comment by lezcano** [comment]
`auto_linear` means that the forward AD is autogenerated. You should look into the formula for the first parameters.

**Comment by lezcano** [comment]
@Thytu any update here?

**Comment by Thytu** [comment]
> @Thytu any update here?

No progress

Tell me if I'm wrong but to fix the fw/bw grad it will require replacing the auto-generated `auto_linear` by a hand defined derivative fn. Is it?

**Comment by Thytu** [comment]
> 

Giving it a try

**Comment by Thytu** [comment]
So if I understand well, the backward of pixel_shuffle, is just pixel_unshuffle. Right?

**Comment by lezcano** [comment]
yep! It looks like pixel_unshuffle may need a similar fix

**Comment by Thytu** [comment]
Sorry for the waiting time, I locked a time slot to work on it this evening :)

**Comment by Thytu** [comment]
> As mentioned in [#85347 (comment)](https://github.com/pytorch/pytorch/pull/85347#issuecomment-1258387365), `auto_linear` is just for forward AD and forward AD just works. You are having issues with backward AD. The backward AD is the formula after the `self:` in the entry for this function in `derivatives.yaml`. It seems that it's not handling well empty inputs.
> 
> Have a go at it, and if you get stuck, ping me and I'll try to give you a hand.

I admit to be stuck on this one...

* Test failing : `test_fn_fwgrad_bwgrad_nn_functional_pixel_shuffle_cpu_float64`
* To reproduce : `python -bb test_ops_gradients.py -v --import-slow-tests --import-disabled-tests -k test_fn_fwgrad_bwgrad_nn_functional_pixel_shuffle_cpu_float64`
* Test exec `pixel_unshuffle_cpu`

**Comment by Thytu** [comment]
Oh shit 🤡

**Comment by lezcano** [comment]
you may want to close this PR and push a new one :)

**Review by malfet — CHANGES_REQUESTED** [review]
Why this PR touches 1000+ files?

**Pull request closed without merging**